### PR TITLE
feat(tianmu): fix invalid write/read issue. (#817)

### PR DESCRIPTION
**Pull request by lujiashun**
## Summary about this PR

Issue Number: ref #817

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [X] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [x] New Feature
- [ ] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/system/cacheable_item.cpp [hunk 1]
```diff
@@ -34,7 +34,7 @@ CacheableItem::CacheableItem(char const *owner_name, char const *object_id, int
     // read the configuration parameter
     std::string temp_filename = tianmu_sysvar_cachefolder;
     filename_n_position = temp_filename.length();
-    filename = new char[filename_n_position + 37];  // "...path.../XXXXXXnnnnnnAAAAAAAABBBBBBBB.tianmu_tmp"
+    filename = new char[filename_n_position + 37 + 4];  // "...path.../XXXXXXnnnnnnAAAAAAAABBBBBBBB.tianmu_tmp"
```
Comment (RingsC) [inline storage/tianmu/system/cacheable_item.cpp]: Pls, would you like use cosntexpr or meaningful words to describe these magic numbers?

File: storage/tianmu/core/engine_results.cpp [hunk 2]
```diff
@@ -36,7 +36,7 @@ void scan_fields(mem_root_deque<Item *> &fields, uint *&buf_lens, std::map<int,
   Item::Type item_type;
   Item_sum::Sumfunctype sum_type;
 
-  buf_lens = new uint[CountVisibleFields(fields)];  // stonedb8
+  buf_lens = new uint[fields.size()];
```
Comment (RingsC) [inline storage/tianmu/core/engine_results.cpp:38]: If we do `memset` after a new memory allocated that would be a good thing.

**Review by RingsC — CHANGES_REQUESTED** [review]

**Review by RingsC — APPROVED** [review]
LGTM

File: storage/tianmu/system/cacheable_item.cpp [hunk 3]
```diff
@@ -23,153 +23,172 @@
...
-      filename_n_position++;
-    }
-    filename_n_position += 6;
-  }
 
-  if (filename == nullptr) {
-    // if the temporary path is not set, use the current folder
-    filename = new char[36];  // "XXXXXXnnnnnnAAAAAAAABBBBBBBB.tianmu_tmp"
-    filename_n_position = 6;
+  constexpr size_t MIN_OWN_NAME_LEN = 3;
```
Comment (hustjieke) [inline storage/tianmu/system/cacheable_item.cpp]: The `constant names` should follow:
https://google.github.io/styleguide/cppguide.html#Constant_Names

File: storage/tianmu/system/cacheable_item.cpp [hunk 4]
```diff
@@ -23,153 +23,172 @@
 
 namespace Tianmu {
 namespace system {
-CacheableItem::CacheableItem(char const *owner_name, char const *object_id, int _default_block_size) {
-  default_block_size = _default_block_size;
+CacheableItem::CacheableItem(char const *owner_name, char const *object_id, int default_block_size) {
+  default_block_size_ = default_block_size;
+  max_file_id_ = 0;
```
Comment (hustjieke) [inline storage/tianmu/system/cacheable_item.cpp]: These class variables can be inited during definition in class. Follow c++11 std.
https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

**Review by hustjieke — CHANGES_REQUESTED** [review]

**Review by hustjieke — APPROVED** [review]
LGTM

**Pull request merged**